drivers/input/input_pinnacle.c

```diff
@@ -0,0 +1,827 @@
+	const struct spi_buf rx_buf[] = {{
+		.buf = rx_data,
+		.len = ARRAY_SIZE(rx_data),
+	}};
+	const struct spi_buf_set rx_set = {
+		.buffers = rx_buf,
+		.count = ARRAY_SIZE(rx_buf),
+	};
+
+	int rc;
+
+	rc = spi_transceive_dt(&bus->spi, &tx_set, &rx_set);
+	if (rc) {
+		LOG_ERR("Failed to read from SPI %s", bus->spi.bus->name);
+		return rc;
+	}
+
+	*value = rx_data[3];
+
+	return 0;
+}
+
+static int pinnacle_seq_read_spi(const struct pinnacle_bus *bus, uint8_t address, uint8_t *buf,
+				 uint8_t count)
+{
+
+	uint8_t size = count + 3;
+	uint8_t tx_data[size];
+
+	tx_data[0] = PINNACLE_READ_REG(address);
+	tx_data[1] = PINNACLE_SPI_FC;
+	tx_data[2] = PINNACLE_SPI_FC;
+
+	uint8_t i = 3;
+
+	for (; i < (count + 2); ++i) {
+		tx_data[i] = PINNACLE_SPI_FC;
+	}
+
+	tx_data[i++] = PINNACLE_SPI_FB;
+
+	const struct spi_buf tx_buf[] = {{
+		.buf = tx_data,
+		.len = size,
+	}};
+	const struct spi_buf_set tx_set = {
+		.buffers = tx_buf,
+		.count = 1,
+	};
+
+	uint8_t rx_data[size];
+	const struct spi_buf rx_buf[] = {{
+		.buf = rx_data,
+		.len = size,
+	}};
+	const struct spi_buf_set rx_set = {
+		.buffers = rx_buf,
+		.count = 1,
+	};
+
+	int rc;
+
+	rc = spi_transceive_dt(&bus->spi, &tx_set, &rx_set);
+	if (rc) {
+		LOG_ERR("Failed to read from SPI %s", bus->spi.bus->name);
+		return rc;
+	}
+
+	for (i = 0; i < count; ++i) {
+		buf[i] = rx_data[i + 3];
+	}
+
+	return 0;
+}
+#endif /* DT_ANY_INST_ON_BUS_STATUS_OKAY(spi) */
+
+static void pinnacle_decode_sample(const struct device *dev, uint8_t *rx,
+				   union pinnacle_sample *sample)
+{
+	const struct pinnacle_config *config = dev->config;
+
+	if (config->data_mode == PINNACLE_DATA_MODE_ABS) {
+		sample->abs_x = ((rx[2] & 0x0F) << 8) | rx[0];
+		sample->abs_y = ((rx[2] & 0xF0) << 4) | rx[1];
+		sample->abs_z = rx[3] & 0x3F;
+	} else if (config->data_mode == PINNACLE_DATA_MODE_REL) {
+		sample->rel_x = ((rx[0] & BIT(4)) == BIT(4)) ? -(256 - rx[1]) : rx[1];
+		sample->rel_y = ((rx[0] & BIT(5)) == BIT(5)) ? -(256 - rx[2]) : rx[2];
+	}
+}
+
+static bool pinnacle_is_idle_sample(const union pinnacle_sample *sample)
+{
+	return (sample->abs_x == 0 && sample->abs_y == 0 && sample->abs_z == 0);
+}
+
+static void pinnacle_clip_sample(const struct device *dev, union pinnacle_sample *sample)
+{
+	const struct pinnacle_config *config = dev->config;
+
+	if (sample->abs_x < config->active_range_x_min) {
+		sample->abs_x = config->active_range_x_min;
+	}
+	if (sample->abs_x > config->active_range_x_max) {
+		sample->abs_x = config->active_range_x_max;
+	}
+	if (sample->abs_y < config->active_range_y_min) {
+		sample->abs_y = config->active_range_y_min;
+	}
+	if (sample->abs_y > config->active_range_y_max) {
+		sample->abs_y = config->active_range_y_max;
+	}
+}
+
+static void pinnacle_scale_sample(const struct device *dev, union pinnacle_sample *sample)
+{
+	const struct pinnacle_config *config = dev->config;
+
+	sample->abs_x = (uint16_t)((uint32_t)(sample->abs_x - config->active_range_x_min) *
```

Comment:
what if `active_range_x_min` is set too high? this would go negative and overflow, would it make more sense to do the computation as `int32_t`, let it go negative and then clamp it down (there's a `CLAMP` macro) to the valid range?

Reply:
This can't be because `pinnacle_scale_sample` is called after `pinnacle_clip_sample` where the latter one guarantees that `sample` absolute coordinates are within the active range.